---
title: "Heatmap 5PSeq microbiome"
author: "Vicent Pelechano"
date: '2018-07-29'
output:
  pdf_document: default
  html_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


```{r Load the cluster data,include=FALSE}
# First I load the required packages
library("Biostrings")
library("ShortRead")
library("LSD")
library("RColorBrewer")
library("DESeq2")# 
library("gplots")
library("ggplot2")
#library("venneuler")
library(sm)
#library(VennDiagram)
#library("lattice")
#library("vioplot")
#library(rtracklayer)
#library(seqLogo) 
#library(BSgenome)
library(ComplexHeatmap)
library(circlize)

serverDir="/Users/pelechano/Dropbox/SciLifeLab/Projects/"


projectDir=file.path(serverDir,"Microbiome_5PSeq/vagaganome/2018-07-30")

dataDir=file.path(projectDir,"data")
dataDir2=file.path(projectDir,"data2")
resDir=file.path(projectDir,"res")
#visuDir=file.path(projectDir,"visu")
setwd<-resDir


# First I read the files deposited in the dataset
files<-list.files(dataDir,full.names=TRUE)

# This has to chnge dependeing of the naming structure
nfiles<-list.files(dataDir)
nfiles2<-sapply(nfiles,function(x) strsplit(x,split="_")[[1]])
anndata<-data.frame(sp=nfiles2[1,],type=nfiles2[2,],A=nfiles2[3,],B=nfiles2[4,],samp=nfiles2[5,],i=1:length(nfiles2[1,]))

#small function to read the data generated by Maryia
readAAprofile<-function(ftoread){
  x<-read.table(ftoread,comment.char=">") # to do not read those lines
  namx<-read.table(ftoread,fill=TRUE)[,1]
  rownames(x)<-namx[seq(1,length(namx),2)]
  colnames(x)<-c(-50:49)
  return(x)
}

# Now I apply in batch to all the data
oprofiles<-lapply(files,function(x) readAAprofile(x))
names(oprofiles)<-nfiles


save(oprofiles,anndata,file=file.path(resDir,"test_profiles_mix.rda"))

```


```{r, echo=FALSE}

# Here I define The columns I want to plot from the 100 span 
poslim_u = 20
poslim_d = 60


plotheat_simple<-function(i){
  subprof<-as.matrix(profiles[[i]][,poslim_u:poslim_d])
  heatmap.2(subprof,distfun=function(x) as.dist((1-cor(t(x),method="pearson"))/2),col=rev(heat.colors(256)),Colv = NA,dendrogram="row",scale="row",trace="none",na.rm=TRUE, main=names(profiles)[i])#colsep=c(17,24,28),sepcolor="darkgrey"

}

# Now I compute wich fraction of the reads correspond to the position -14 [37]
# Function to do that computation
getvector14<-function(x){
    posofinterest<-37 # corresponds to -14
  subprof<-(x[,poslim_u:poslim_d])
  outvect<-x[,posofinterest]/rowSums(subprof)
  return(outvect)
}



# Define data to test

sp="ATCC8014"
type="aminoacids"


plot_mult_heat<-function(sp,type) {
totest<-subset(anndata$i,anndata$type==type & anndata$sp==sp)
profiles<-oprofiles[totest]
# Plot each sample
#sapply(totest,function(x) plotheat_simple(x))
pdf(file.path(resDir,paste(sp,"_",type,".pdf",sep="")),10,10)
for(j in 1:length(totest)){#print(j)
  plotheat_simple(j) }
dev.off()

###
testout<-lapply(profiles,function(x) getvector14(x))
testout<-as.data.frame(testout)
names(testout)<-paste(anndata$A,anndata$B,anndata$samp,sep="_")[totest]

pdf(file.path(resDir,paste("comb_",sp,"_",type,".pdf",sep="")),10,10)
heatmap.2(as.matrix(testout),distfun=function(x) as.dist((1-cor(t(x),method="pearson"))/2), col=rev(heat.colors(256)),Rowv = NA,dendrogram="col",scale="col",trace="none",na.rm=TRUE, main=strsplit(names(profiles)[1],split="_")[[1]][1])#colsep=c(17,24,28),sepcolor="darkgrey"
dev.off()
}


plot_mult_heat(sp="ATCC8014",type="aminoacids") 
plot_mult_heat(sp="DSM",type="aminoacids") 

```
 
 
 
 
 Now I repeat the same for the not mixed samples
 
 
 
```{r, echo=FALSE}

# First I read the files deposited in the dataset
files<-list.files(dataDir2,full.names=TRUE)

# This has to chnge dependeing of the naming structure
nfiles<-list.files(dataDir2)
nfiles2<-sapply(nfiles,function(x) strsplit(x,split="_")[[1]])
anndata<-data.frame(type=nfiles2[1,],A=nfiles2[2,],B=nfiles2[3,],samp=nfiles2[4,],i=1:length(nfiles2[1,]))


# Now I apply in batch to all the data
oprofiles<-lapply(files,function(x) readAAprofile(x))
names(oprofiles)<-nfiles


save(oprofiles,anndata,file=file.path(resDir,"test_profiles_notmixed.rda"))

# Here I define The columns I want to plot from the 100 span 
poslim_u = 20
poslim_d = 60


# Define data to test

type="aminoacids"


plot_mult_heat2<-function(type) {
totest<-subset(anndata$i,anndata$type==type)
profiles<-oprofiles[totest]
# Plot each sample
#sapply(totest,function(x) plotheat_simple(x))
pdf(file.path(resDir,paste(type,".pdf",sep="")),10,10)
for(j in 1:length(totest)){
  #print(j)
  plotheat_simple(j) }
dev.off()

###
testout<-lapply(profiles,function(x) getvector14(x))
testout<-as.data.frame(testout)
names(testout)<-paste(anndata$A,anndata$B,anndata$samp,sep="_")[totest]

pdf(file.path(resDir,paste("comb_",type,".pdf",sep="")),10,10)
heatmap.2(as.matrix(testout),distfun=function(x) as.dist((1-cor(t(x),method="pearson"))/2), col=rev(heat.colors(256)),Rowv = NA,dendrogram="col",scale="col",trace="none",na.rm=TRUE, main=strsplit(names(profiles)[1],split="_")[[1]][1])#colsep=c(17,24,28),sepcolor="darkgrey"
dev.off()
}


plot_mult_heat2(type="aminoacids") 
#plot_mult_heat2(type="codons") 


```
 
 
 
 
 
 
 
 
 
 
 
 
 
For the Future
```{r, echo=FALSE}
# 
# ht1 = Heatmap(matDE,show_row_dend = FALSE,name = "DE", clustering_distance_rows = "pearson",
#         col = colorRamp2(c(-5, 0, 5), c("lightblue", "white", "red")),na_col = "grey",
#         row_title = "Analysed TSSs",column_title = "Strains")
# ha_row2 = rowAnnotation(df = data.frame(intCDS = intCDS), col = list(intCDS = c("iTSS" =  "black", "No" = "white")))
# 
# ha_row = rowAnnotation(df = matDEup, col = list(DE_up_rrp6 = c("A" =  "red", "B" = "lightgrey"),
#                        DE_up_set1 = c("A" =  "red", "B" = "lightgrey"),
#                        DE_up_eaf3 = c("A" =  "red", "B" = "lightgrey"),
#                        DE_up_set2 = c("A" =  "red", "B" = "lightgrey"),
#                        DE_up_rco1 = c("A" =  "red", "B" = "lightgrey")))
# 
# ht_list = ht1 + ha_row2 + ha_row
# draw(ht_list)


```

